crates/nu-path/src/pwd_per_drive.rs

```diff
@@ -0,0 +1,349 @@
+use std::path::{Path, PathBuf};
+
+#[derive(PartialEq, Debug)]
+pub enum PathError {
+    InvalidDriveLetter,
+    InvalidPath,
+    CantLockSharedMap,
+}
+/// Usage for pwd_per_drive on windows
+///
+/// Upon change PWD, call set_pwd() with absolute path
+///
+/// Call expand_pwd() with relative path to get absolution path
+///
+/// ```
+/// use std::path::{Path, PathBuf};
+/// use nu_path::{expand_pwd, set_pwd};
+///
+/// // Set PWD for drive C
+/// set_pwd(Path::new(r"C:\Users\Home")).unwrap();
+///
+/// // Expand a relative path
+/// let expanded = expand_pwd(Path::new("c:test"));
+/// assert_eq!(expanded, Some(PathBuf::from(r"C:\Users\Home\test")));
+///
+/// // Will NOT expand an absolute path
+/// let expanded = expand_pwd(Path::new(r"C:\absolute\path"));
+/// assert_eq!(expanded, None);
+///
+/// // Expand with no drive letter
+/// let expanded = expand_pwd(Path::new(r"\no_drive"));
+/// assert_eq!(expanded, None);
+///
+/// // Expand with no PWD set for the drive
+/// let expanded = expand_pwd(Path::new("D:test"));
+/// assert!(expanded.is_some());
+/// let abs_path = expanded.unwrap().as_path().to_str().expect("OK").to_string();
+/// assert!(abs_path.starts_with(r"D:\"));
+/// assert!(abs_path.ends_with(r"\test"));
+/// ```
+pub mod singleton {
+    use super::*;
+    use crate::pwd_per_drive::PathError::CantLockSharedMap;
+
+    /// set_pwd_per_drive
+    /// Record PWD for drive, path must be absolute path
+    /// return Ok(()) if succeeded, otherwise error message
+    pub fn set_pwd(path: &Path) -> Result<(), PathError> {
+        if let Ok(mut pwd_per_drive) = get_drive_pwd_map().lock() {
+            pwd_per_drive.set_pwd(path)
+        } else {
+            Err(CantLockSharedMap)
+        }
+    }
+
+    /// expand_pwe_per_drive
+    /// Input relative path, expand PWD-per-drive to construct absolute path
+    /// return PathBuf for absolute path, None if input path is invalid.
+    pub fn expand_pwd(path: &Path) -> Option<PathBuf> {
+        if need_expand_pwd_per_drive(path) {
+            if let Ok(mut pwd_per_drive) = get_drive_pwd_map().lock() {
+                return pwd_per_drive.expand_path(path);
+            }
+        }
+        None
+    }
+}
+
+/// Helper to check if input path is relative path
+/// with drive letter, it can be expanded with PWD-per-drive.
+fn need_expand_pwd_per_drive(path: &Path) -> bool {
+    if let Some(path_str) = path.to_str() {
+        let chars: Vec<char> = path_str.chars().collect();
+        if chars.len() >= 2 {
+            return chars[1] == ':' && (chars.len() == 2 || (chars[2] != '/' && chars[2] != '\\'));
+        }
+    }
+    false
+}
+
+#[test]
```

Comment:
can you please put this test at the bottom of the file with all the other tests?

Reply:
OK